### PR TITLE
Added Changes for LiveQuery Options

### PR DESCRIPTION
**Pull request by AliabbasMerchant**
@YourTechBud Done.  
This replaces the previous PR. (Which I have closed)

### CODE REVIEW

**Comment by YourTechBud** [comment]
@AliabbasMerchant is this complete? You still seem to be firing the get request internally event when initial is supposed to be neglected. Can we avoid that query?

**Comment by YourTechBud** [comment]
@AliabbasMerchant we can't use `AddLiveQuery` directly. That bypasses the entire security module.

**Comment by AliabbasMerchant** [comment]
@YourTechBud Please have a look. I have tested it and is working.

File: modules/realtime/realtime.go [hunk 1]
```diff
@@ -54,6 +54,10 @@ type pendingRequest struct {
 	time time.Time
 }
 
+type LiveQueryOptions struct {
```
Comment (YourTechBud) [inline modules/realtime/realtime.go]: Define this in the `model` package

File: model/realtime.go [hunk 2]
```diff
@@ -20,6 +20,7 @@ type RealtimeRequest struct {
 	Type    string                 `json:"type"`  // Can either be subscribe or unsubscribe
 	ID      string                 `json:"id"`    // id is the query id
 	Where   map[string]interface{} `json:"where"`
+	Options map[string]interface{} `json:"options"`
```
Comment (YourTechBud) [inline model/realtime.go]: Keep this of type `LiveQueryOptions`

File: modules/realtime/operations.go [hunk 3]
```diff
@@ -21,6 +23,14 @@ func (m *Module) Subscribe(ctx context.Context, clientID string, auth *auth.Modu
 		return nil, err
 	}
 
+	options := new(LiveQueryOptions)
+	mapstructure.Decode(data.Options, data)
+
+	if options.SkipInitial {
+		m.AddLiveQuery(data.ID, data.Project, data.Group, clientID, data.Where, sendFeed)
+		return nil, nil
```
Comment (YourTechBud) [inline modules/realtime/operations.go]: Lets return an empty array of feed data?

File: utils/client/grpc_realtime.go [hunk 4]
```diff
@@ -90,6 +90,12 @@ func (c *GRPCRealtimeClient) Read(cb DataCallback) {
 			return
 		}
 		data["Where"] = temp
+		err = json.Unmarshal(in.Options, &temp)
```
Comment (YourTechBud) [inline utils/client/grpc_realtime.go]: Don't forget to clear temp before this step. Re initialise temp.

Comment (YourTechBud) [reply utils/client/grpc_realtime.go]: Temp should be of type LiveQueryOptions i guess

File: utils/server/grpc.go [hunk 5]
```diff
@@ -408,8 +409,13 @@ func (s *Server) RealTime(stream pb.SpaceCloud_RealTimeServer) error {
 				return true
 			}
 
+			options := new(realtime.LiveQueryOptions)
```
Comment (YourTechBud) [inline utils/server/grpc.go]: Lets remove these steps complete. Lets send the empty feed data array as is. On the client we can ignore the message if feed data array is empty

File: utils/server/websocket.go [hunk 6]
```diff
@@ -54,8 +55,13 @@ func (s *Server) handleWebsocket() http.HandlerFunc {
 					return true
 				}
 
+				options := new(realtime.LiveQueryOptions)
```
Comment (YourTechBud) [inline utils/server/websocket.go]: This part is not required

**Comment by AliabbasMerchant** [comment]
@YourTechBud Perfect! Your requested changes made the logic quite easy to understand.  
I have made the changes.  
This the requested changes do not break any of the clients and do not change the client implementation.

File: model/realtime.go [hunk 7]
```diff
@@ -20,6 +20,7 @@ type RealtimeRequest struct {
 	Type    string                 `json:"type"`  // Can either be subscribe or unsubscribe
 	ID      string                 `json:"id"`    // id is the query id
 	Where   map[string]interface{} `json:"where"`
+	Options LiveQueryOptions       `json:"options"`
```
Comment (YourTechBud) [inline model/realtime.go]: LiveQueryOptions must be a pointer. Prefix it with *

File: model/realtime.go [hunk 8]
```diff
@@ -37,3 +38,7 @@ type Message struct {
 	Data interface{} `json:"data"`
 	ID   string      `json:"id"` // the request id
 }
+
+type LiveQueryOptions struct {
```
Comment (YourTechBud) [inline model/realtime.go]: Add the export comment

File: model/realtime.go [hunk 9]
```diff
@@ -37,3 +38,7 @@ type Message struct {
...
+	SkipInitial bool `json:"skipInitial"`
```
Comment (YourTechBud) [inline model/realtime.go]: Wasn't this supposed to be change only. Lets use a consistent nomenclature across

File: modules/realtime/operations.go [hunk 10]
```diff
@@ -21,6 +23,14 @@ func (m *Module) Subscribe(ctx context.Context, clientID string, auth *auth.Modu
 		return nil, err
 	}
 
+	options := new(model.LiveQueryOptions)
+	mapstructure.Decode(data.Options, options)
```
Comment (YourTechBud) [inline modules/realtime/operations.go]: Why do you need this step?

File: model/realtime.go (hunk 9 above)
Comment (AliabbasMerchant) [reply model/realtime.go]: I think no. They are 2 diff things. The server has nothing to do with `changesOnly`. That is handled entirely on the client. The server just needs to know whether to send the initial or not. (Internally, in the clients, I am setting the values of both to be the same)

File: modules/realtime/operations.go (hunk 10 above)
Comment (AliabbasMerchant) [reply modules/realtime/operations.go]: Oh. Right

File: model/realtime.go (hunk 9 above)
Comment (YourTechBud) [reply model/realtime.go]: Cool. Works for me.

**Pull request merged**